src/test/opcat_tests.cpp

```diff
@@ -0,0 +1,142 @@
+    TaprootSpendData spendData = builder.GetSpendData();
+    CAmount nValue = 1;
+    const CTransaction txCredit{BuildCreditingTransaction(scriptPubkey, nValue)};
+
+    CScript scriptSig = CScript(); // Script sig is only size 0 and empty in tapscript
+    CMutableTransaction txSpend = BuildSpendingTransaction(scriptSig, CScriptWitness(), txCredit);
+    CScriptWitness& scriptWitness = txSpend.vin[0].scriptWitness;
+
+    auto controlblocks = spendData.scripts[{witVerifyScript, TAPROOT_LEAF_TAPSCRIPT}];
+    auto controlblock = *(controlblocks.begin());
+
+    for (auto witDatum : witData) {
+        scriptWitness.stack.push_back(witDatum);
+    }
+    scriptWitness.stack.push_back(witVerifyScript);
+    scriptWitness.stack.push_back(controlblock);
+
+    // unsigned int flags = 0;
+    unsigned int flags = SCRIPT_VERIFY_P2SH;
+    flags += SCRIPT_VERIFY_WITNESS;
+    flags += SCRIPT_VERIFY_TAPROOT;
+    flags += SCRIPT_VERIFY_TAPSCRIPT_OP_CAT;
+    ScriptError err;
+
+    bool success = VerifyScript(
+        txSpend.vin[0].scriptSig,
+        txCredit.vout[0].scriptPubKey,
+        &txSpend.vin[0].scriptWitness,
+        flags,
+        MutableTransactionSignatureChecker(&txSpend, 0, txCredit.vout[0].nValue, MissingDataBehavior::ASSERT_FAIL),
+        &err);
+
+    BOOST_CHECK_MESSAGE(success == expect, message);
+    BOOST_CHECK_MESSAGE(err == scriptError, std::to_string(err) + " where " + std::to_string(scriptError) + " expected: " + message);
+}
+
+BOOST_FIXTURE_TEST_SUITE(opcat_tests, BasicTestingSetup)
+
+// Most of the OP_CAT tests live in data/script_tests.json
+
+BOOST_AUTO_TEST_CASE(cat_simple)
+{
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back(ParseHex("aa"));
+    witData.push_back(ParseHex("bbbb"));
+
+    std::vector<unsigned char> witVerifyScript = { OP_CAT,
+    OP_PUSHDATA1, 0x03, 0xaa, 0xbb, 0xbb,
+    OP_EQUAL };
+
+    DoTest(witVerifyScript, witData, SCRIPT_ERR_OK, "cat_simple");
+}
+
+
+BOOST_AUTO_TEST_CASE(cat_empty_stack)
+{
+    // Ensures that OP_CAT successfully handles concatenating two empty stack elements
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back({});
+    witData.push_back({});
+
+    std::vector<unsigned char> witVerifyScript = 
+    { OP_CAT, 
+    OP_PUSHDATA1, 0x00,
+    OP_EQUAL };
+
+    DoTest(witVerifyScript, witData, SCRIPT_ERR_OK, "cat_empty_stack");
+}
+
+BOOST_AUTO_TEST_CASE(cat_dup_test)
+{
+    // CAT DUP exhaustion attacks using all element sizes from 1 to 522
+    // with CAT DUP repetitions up to 10. Ensures the correct error is thrown
+    // or not thrown as appropriate. 
+    unsigned int maxElementSize = 522;
+    unsigned int maxDupsToCheck = 10;
+
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back({});
+
+    for (unsigned int elementSize = 1; elementSize <= maxElementSize; elementSize++) {
+        std::vector<unsigned char> witVerifyScript;
+        // increase the size of stack element by one byte
```

Comment:
Would this increase the stack element size by one or add a new stack element altogether?

Reply:
This increases the size of the stack element, it does not increase the number of stack elements.